TESTS/block_device/qspif/main.cpp

```diff
@@ -0,0 +1,339 @@
+    _mutex->lock();
+    srand(seed);
+    for (bd_size_t i_ind = 0; i_ind < block_size; i_ind++) {
+        write_block[i_ind] = 0xff & rand();
+    }
+    // Write, sync, and read the block
+    printf("\ntest  %0*llx:%llu...", addrwidth, block, block_size);
+    _mutex->unlock();
+
+    err = blockD.erase(block, block_size);
+    TEST_ASSERT_EQUAL(0, err);
+
+    err = blockD.program(write_block, block, block_size);
+    TEST_ASSERT_EQUAL(0, err);
+
+    _mutex->lock();
+    printf("\nwrite %0*llx:%llu ", addrwidth, block, block_size);
+    for (int i_ind = 0; i_ind < 16; i_ind++) {
+        printf("%02x", write_block[i_ind]);
+    }
+    printf("...\n");
+    _mutex->unlock();
+
+    err = blockD.read(read_block, block, block_size);
+    TEST_ASSERT_EQUAL(0, err);
+
+    _mutex->lock();
+    printf("read  %0*llx:%llu ", addrwidth, block, block_size);
+    for (int i_ind = 0; i_ind < 16; i_ind++) {
+        printf("%02x", read_block[i_ind]);
+    }
+    printf("...\n");
+    _mutex->unlock();
+
+    // Find error mask for debugging
+    memset(error_mask, 0, TEST_ERROR_MASK);
+    bd_size_t error_scale = block_size / (TEST_ERROR_MASK * 8);
+
+    _mutex->lock();
+    srand(seed);
+    for (bd_size_t i_ind = 0; i_ind < TEST_ERROR_MASK * 8; i_ind++) {
+        for (bd_size_t j_ind = 0; j_ind < error_scale; j_ind++) {
+            if ((0xff & rand()) != read_block[i_ind * error_scale + j_ind]) {
+                error_mask[i_ind / 8] |= 1 << (i_ind % 8);
+            }
+        }
+    }
+
+    printf("error %0*llx:%llu ", addrwidth, block, block_size);
+    for (int i_ind = 0; i_ind < 16; i_ind++) {
+        printf("%02x", error_mask[i_ind]);
+    }
+    printf("\n");
+
+    // Check that the data was unmodified
+    srand(seed);
+    for (bd_size_t i_ind = 0; i_ind < block_size; i_ind++) {
+        TEST_ASSERT_EQUAL(0xff & rand(), read_block[i_ind]);
+    }
+    _mutex->unlock();
+}
+
+void test_qspif_random_program_read_erase()
+{
+
+    printf("\nTest Random Program Read Erase Starts..\n");
+
+    QSPIFBlockDevice blockD(MBED_CONF_QSPIF_QSPI_IO0, MBED_CONF_QSPIF_QSPI_IO1, MBED_CONF_QSPIF_QSPI_IO2,
+                            MBED_CONF_QSPIF_QSPI_IO3, MBED_CONF_QSPIF_QSPI_CLK, MBED_CONF_QSPIF_QSPI_CS, QSPIF_POLARITY_MODE_0,
+                            MBED_CONF_QSPIF_QSPI_FREQ);
+
+    int err = blockD.init();
+    TEST_ASSERT_EQUAL(0, err);
+
+    for (unsigned atr = 0; atr < sizeof(ATTRS) / sizeof(ATTRS[0]); atr++) {
+        static const char *prefixes[] = {"", "k", "M", "G"};
+        for (int i_ind = 3; i_ind >= 0; i_ind--) {
+            bd_size_t size = (blockD.*ATTRS[atr].method)();
+            if (size >= (1ULL << 10 * i_ind)) {
+                printf("%s: %llu%sbytes (%llubytes)\n",
+                       ATTRS[atr].name, size >> 10 * i_ind, prefixes[i_ind], size);
+                break;
+            }
+        }
+    }
+
+    bd_size_t block_size = blockD.get_erase_size();
+    uint8_t *write_block = new uint8_t[block_size];
```

Comment:
I don't know how large this block size can get. If it's large, the test can fail on low memory boards. In this case, would suggest to add `std::nothrow` to the allocation. [This](https://github.com/ARMmbed/mbed-os/pull/7465/files#diff-4e5d26193a7c8cf8e05e95d7775dcf84R76) can be used as a good reference.

Reply:
It can get quite large (Typically 4KB) - I'll add 'std::nothrow'  - thanks